File: packages/react-native/ios/C15tReactNative/ReactNative/C15tReactNativeModule.swift
#if canImport(C15tCore)
// SwiftPM builds the consent kernel as its own module. CocoaPods compiles the vendored
// copy into this module instead, where the import would fail, so it is conditional
// rather than assumed.
import C15tCore
#endif
#if canImport(C15tReactNativeBridge)
// SwiftPM builds the wire layer as its own module. CocoaPods compiles it into this
// module instead, where the import would fail, so it is conditional rather than
// assumed.
import C15tReactNativeBridge
#endif
import Foundation
import React

/// The `C15t` TurboModule: the JavaScript surface in `src/specs/NativeC15t.ts`,
/// forwarded call for call to the Swift consent core.
///
/// Nothing in here decides anything about consent. It reads state the core already
/// holds, forwards the actions the user took, and reports the results as the JSON the
/// JavaScript protocol expects. That is deliberate: there is one consent kernel, and a
/// rule repeated here would be a second one to keep honest. The behaviour lives in
/// `C15tModuleHandler`, which has no React Native in it.
///
/// Threading. ``getBootstrap()`` and ``getSnapshot()`` are synchronous methods and do
/// no disk or network work: the core keeps one immutable snapshot in memory, so a read
/// is a lock plus an encode. They never hop to a core queue and wait, which is the
/// deadlock the New Architecture warns about for a synchronous module method, and ad
/// SDKs call both on the main thread. Everything that could block returns a promise,
/// and those promises resolve as soon as the local commit is durable, with delivery
/// left to the core's own scheduler.
///
/// The class deliberately omits the `NativeC15tSpec` conformance that the TypeScript
/// spec implies. Codegen emits that protocol into an ObjC++ umbrella header whose first
/// line is `#error This file must be compiled as Obj-C++`, so no Swift file can name it:
/// the conformance, and the `getTurboModule:` factory `RCTModuleProviders` requires, live
/// in `C15tReactNativeModule.mm`.
///
/// That split leaves one sharp edge. The generated JSI glue reaches this class by
/// `@selector`, and a selector that disagrees with the protocol still compiles, so every
/// method below states its Objective-C name rather than relying on inference, and
/// `src/specs/__tests__/ios-spec-surface.test.ts` fails the moment those names and the
/// protocol Codegen generates stop agreeing.
///
@objc(C15tReactNativeModule)
public final class C15tReactNativeModule: RCTEventEmitter {
    private let handler = C15tModuleHandler()

    /// Held as a property: the core keeps snapshot observers weakly, so somebody has
    /// to own the observer or the callbacks would silently stop.
    private lazy var subscriber = C15tCoreSubscriber(sink: C15tReactEmitterSink(module: self))

    /// Live JavaScript listeners. A snapshot event is only a hint to go and read, so
    /// with no subscriber it is dropped rather than queued.
    private var listenerCount = 0

    public override init() {
        super.init()
        followCore()
    }

    @available(*, unavailable)
    required public init?(bridge: RCTBridge) {
        fatalError("init(bridge:) is unavailable: this module is New Architecture only")
    }

    // MARK: - Synchronous reads

    /// Handshake payload: the cached snapshot, never a network wait.
    @objc(getBootstrap)
    public func getBootstrap() -> String {
        followCore()
        return handler.bootstrapPayload()
    }

    /// The current snapshot, encoded for the wire.
    @objc(getSnapshot)
    public func getSnapshot() -> String {
        followCore()
        return handler.snapshotPayload()
    }

    /// The platform tracking answer.
    ///
    /// No `followCore()`: this reads Apple, not the consent core, and the answer is the
    /// same whether or not a core was ever started. It is also not a consent read, and a
    /// host that treats it as one gets a category the subject never granted.
    @objc(getTrackingAuthorization)
    public func getTrackingAuthorization() -> String {
        handler.trackingAuthorizationPayload()
    }

    // MARK: - Async commands

    @objc(commit:resolve:reject:)
    public func commit(
        _ intent: String,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        followCore()
        resolve(handler.commitPayload(intent: intent))
    }

    @objc(setOverrides:resolve:reject:)
    public func setOverrides(
        _ overrides: String,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        switch handler.applyOverrides(overrides) {
        case .success:
            resolve(nil)
        case let .failure(error):
            reject(error.code, error.message, nil)
        }
    }

    /// Record that the notice was dismissed. Local and synchronous: it never writes to
    /// the backend, so there is nothing to reject.
    @objc(dismissNotice)
    public func dismissNotice() {
        followCore()
        handler.dismissNotice()
    }

    @objc(refresh:reject:)
    public func refresh(
        _ resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        switch handler.refreshAll() {
        case .success:
            resolve(nil)
        case let .failure(error):
            reject(error.code, error.message, nil)
        }
    }

    @objc(identify:resolve:reject:)
    public func identify(
        _ externalId: String,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        switch handler.identify(externalId: externalId) {
        case .success:
            resolve(nil)
        case let .failure(error):
            reject(error.code, error.message, nil)
        }
    }

    /// Ask Apple for tracking authorization.
    ///
    /// Nothing in this module calls it, and nothing else in the package does: the prompt
    /// belongs after the host's own consent UI, so that the system dialog is never the
    /// first thing a subject reads about tracking. It rejects when the build carries no
    /// prompt string, because Apple would then show nothing and spend the install's one
    /// dialog on a `denied` nobody chose.
    ///
    /// Resolves off Apple's callback, which is not the calling thread. Rationale: the
    /// promise blocks dispatch, and the alternative is a second hop that could queue behind
    /// the presentation this is waiting on.
    @objc(requestTrackingAuthorization:reject:)
    public func requestTrackingAuthorization(
        _ resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        handler.requestTrackingAuthorization { result in
            switch result {
            case let .success(requestResult):
                resolve(C15tPayload.trackingRequestResult(requestResult))
            case let .failure(error):
                reject(error.code, error.message, nil)
            }
        }
    }

    @objc(logout:reject:)
    public func logout(
        _ resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        switch handler.logout() {
        case .success:
            resolve(nil)
        case let .failure(error):
            reject(error.code, error.message, nil)
        }
    }

    @objc(reset:reject:)
    public func reset(
        _ resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        switch handler.reset() {
        case .success:
            resolve(nil)
        case let .failure(error):
            reject(error.code, error.message, nil)
        }
    }

    // MARK: - Events

    public override func supportedEvents() -> [String]! {
        [
            C15tPayload.eventSnapshot,
            C15tPayload.eventError,
            C15tPayload.eventInitialized,
        ]
    }

    @objc(addListener:)
    public override func addListener(_ eventName: String) {
        super.addListener(eventName)
        listenerCount += 1
        followCore()
    }

    @objc(removeListeners:)
    public override func removeListeners(_ count: Double) {
        super.removeListeners(count)
        listenerCount = max(0, listenerCount - Int(count))
    }

    /// Send one event to JavaScript.
    ///
    /// Called from the core's queues, and `RCTEventEmitter` hands the payload to the
    /// dispatcher, so no main-thread hop is needed here.
    fileprivate func publish(event: String, payload: String) {
        guard listenerCount > 0 else { return }
        sendEvent(withName: event, body: payload)
    }

    // MARK: - Lifecycle

    public override static func requiresMainQueueSetup() -> Bool {
        false
    }

    public override func invalidate() {
        subscriber.detach()
        super.invalidate()
    }

    /// Start the core if no launch hook did, and follow it.
    ///
    /// Both are idempotent. A host that set the opt-out flag and never installed a
    /// core keeps getting deny-all answers, which is the contract's answer for state
    /// that does not exist.
    private func followCore() {
        guard handler.ensureCore(), let core = C15t.current else { return }
        subscriber.attach(to: core)
    }
}

/// Bridges the pump to `RCTEventEmitter`, holding the module weakly so the observer
/// the core retains does not keep a torn-down module alive.
private final class C15tReactEmitterSink: C15tEventSink, @unchecked Sendable {
    private weak var module: C15tReactNativeModule?

    init(module: C15tReactNativeModule) {
        self.module = module
    }

    func emit(event: String, payload: String) {
        module?.publish(event: event, payload: payload)
    }
}

File: packages/react-native/android/c15t-react-native/src/main/kotlin/com/c15t/reactnative/C15tReactNativeModule.kt
package com.c15t.reactnative

import com.c15t.core.C15t
import com.c15t.core.Subscription
import com.c15t.core.model.ConsentSnapshot
import com.c15t.core.model.KernelError
import com.c15t.core.model.KernelUser
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.util.Locale

/**
 * The `C15t` TurboModule: the JavaScript surface in
 * `src/specs/NativeC15t.ts`, forwarding call for call to the Kotlin core.
 *
 * Nothing in here decides anything about consent. It reads state the core already
 * holds, forwards the actions the user took, and reports the results as the JSON the
 * JavaScript protocol expects. That is deliberate: there is one consent kernel, and a
 * rule duplicated here would be a second one to keep honest.
 *
 * Threading. [getBootstrap] and [getSnapshot] are synchronous methods and do no disk
 * or network work at all: the core keeps one immutable snapshot in an atomic
 * reference, so a read is a volatile load plus an encode. They never post to the
 * core's worker and wait, which is the deadlock the New Architecture rules warn about
 * for a synchronous module method. Everything that could block returns a promise, and
 * the core's own calls return as soon as the local commit is durable, with delivery
 * left to the core's dispatcher, so the promises resolve on the calling thread
 * without an extra hop that could queue behind the very work being awaited.
 */
class C15tReactNativeModule(appContext: ReactApplicationContext) : NativeC15tSpec(appContext) {
	private val context: ReactApplicationContext = appContext
	private val pump = C15tChangePump(sink = this::publish)

	// Held as fields on purpose: the core keeps snapshot observers weakly, so the
	// module has to own the lambdas or the callbacks would silently stop.
	private val snapshotObserver: (ConsentSnapshot) -> Unit = pump::onSnapshot

	private val errorObserver: (KernelError) -> Unit = pump::onError

	private var snapshotSubscription: Subscription? = null

	private var errorSubscription: Subscription? = null

	@Volatile
	private var listenerCount = 0

	init {
		subscribeToCore()
	}

	/** Handshake payload: the cached snapshot, never a network wait. */
	override fun getBootstrap(): String {
		ensureCore()
		return C15tPayload.bootstrap(
			snapshot = C15t.snapshot(),
			hasStoredSnapshot = C15t.hasStoredSnapshot,
		)
	}

	/** The current snapshot, encoded for the wire. */
	override fun getSnapshot(): String {
		ensureCore()
		return C15tPayload.snapshot(C15t.snapshot(), fallbackLanguage = deviceLanguage())
	}

	override fun commit(
		intent: String,
		promise: Promise,
	) {
		val parsed = C15tPayload.parseCommitIntent(intent)
		if (parsed == null) {
			promise.resolve(C15tPayload.invalidIntent(intent))
			return
		}
		if (!ensureCore()) {
			promise.resolve(C15tPayload.notBootstrapped())
			return
		}
		try {
			val result = C15t.save(parsed)
			promise.resolve(C15tPayload.commitResult(result, C15t.snapshot()))
		} catch (error: Exception) {
			promise.reject(REJECT_COMMIT, error.message, error)
		}
	}

	override fun setOverrides(
		overrides: String,
		promise: Promise,
	) {
		if (!ensureCore()) {
			promise.reject(REJECT_NOT_BOOTSTRAPPED, C15tPayload.MESSAGE_NOT_BOOTSTRAPPED)
			return
		}
		// Replaced rather than merged, because the protocol distinguishes an explicit
		// null from an omitted field and the kernel's own merge cannot. A refusal
		// leaves the live overrides exactly as they were.
		when (val read = C15tPayload.parseOverrides(overrides, C15t.snapshot().overrides)) {
			is OverridesRead.Refused -> promise.reject(read.code, read.message)
			is OverridesRead.Applied -> try {
				C15t.setOverrides(read.overrides, merge = false)
				promise.resolve(null)
			} catch (error: Exception) {
				promise.reject(REJECT_OVERRIDES, error.message, error)
			}
		}
	}

	override fun dismissNotice() {
		if (ensureCore()) {
			C15t.dismissNotice()
		}
	}

	override fun refresh(promise: Promise) {
		if (!ensureCore()) {
			promise.reject(REJECT_NOT_BOOTSTRAPPED, C15tPayload.MESSAGE_NOT_BOOTSTRAPPED)
			return
		}
		try {
			C15t.refresh()
			// The contract's refresh covers the queue too, and the core's refresh only
			// re-runs init.
			C15t.flushPending()
			promise.resolve(null)
		} catch (error: Exception) {
			promise.reject(REJECT_REFRESH, error.message, error)
		}
	}

	override fun identify(
		externalId: String,
		promise: Promise,
	) {
		if (!ensureCore()) {
			promise.reject(REJECT_NOT_BOOTSTRAPPED, C15tPayload.MESSAGE_NOT_BOOTSTRAPPED)
			return
		}
		try {
			C15t.identify(KernelUser(externalId = externalId))
			promise.resolve(null)
		} catch (error: Exception) {
			promise.reject(REJECT_IDENTIFY, error.message, error)
		}
	}

	override fun logout(promise: Promise) {
		if (!ensureCore()) {
			// Sign-out is not a no-op the app can be told about after the fact: with no
			// core there is no subject to detach, and resolving would report a logout
			// that did not happen.
			promise.reject(REJECT_NOT_BOOTSTRAPPED, C15tPayload.MESSAGE_NOT_BOOTSTRAPPED)
			return
		}
		try {
			C15t.logout()
			promise.resolve(null)
		} catch (error: Exception) {
			promise.reject(REJECT_LOGOUT, error.message, error)
		}
	}

	override fun addListener(eventName: String) {
		listenerCount += 1
	}

	override fun removeListeners(count: Double) {
		listenerCount = (listenerCount - count.toInt()).coerceAtLeast(0)
	}

	/** Drop the core subscriptions when React tears this module down. */
	override fun invalidate() {
		snapshotSubscription?.close()
		errorSubscription?.close()
		snapshotSubscription = null
		errorSubscription = null
		super.invalidate()
	}

	/**
	 * Start the core when a launch hook did not.
	 *
	 * @return `true` when a core is installed and reads can mean something.
	 */
	private fun ensureCore(): Boolean {
		val started = C15tReactNativeBootstrap.ensure(context)
		if (started) {
			subscribeToCore()
		}
		return started
	}

	/** Subscribe once the core exists, since a subscription before it would be inert. */
	private fun subscribeToCore() {
		if (C15t.current == null || snapshotSubscription != null) {
			return
		}
		synchronized(this) {
			if (snapshotSubscription != null) {
				return
			}
			snapshotSubscription = C15t.onChange(snapshotObserver)
			errorSubscription = C15t.onError(errorObserver)
		}
		// Anything the core published between launch and this subscription is not lost:
		// JavaScript pulls the snapshot on mount.
		pump.onSnapshot(C15t.snapshot())
	}

	private fun publish(
		event: String,
		payload: String,
	) {
		if (listenerCount == 0) {
			// No subscriber, and a snapshot event is only ever a hint to go and read.
			return
		}
		try {
			context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
				?.emit(event, payload)
		} catch (error: Exception) {
			// React is going away. The next mount pulls the snapshot anyway.
		}
	}

	private fun deviceLanguage(): String = Locale.getDefault().language
		.takeIf { it.isNotBlank() }
		?: C15tPayload.DEFAULT_LANGUAGE

	private companion object {
		const val REJECT_COMMIT = "C15T_COMMIT_FAILED"

		const val REJECT_OVERRIDES = C15tPayload.REJECT_OVERRIDES
		const val REJECT_REFRESH = "C15T_REFRESH_FAILED"
		const val REJECT_IDENTIFY = "C15T_IDENTIFY_FAILED"
		const val REJECT_LOGOUT = "C15T_LOGOUT_FAILED"
		const val REJECT_NOT_BOOTSTRAPPED = "C15T_NOT_BOOTSTRAPPED"
	}
}
